certora/specs/DoubleEndedQueue.spec

```
@@ -0,0 +1,361 @@
+        (f.selector == popBack().selector  && emptyBefore) ||
+        f.selector == at_(uint256).selector
+    ), "only revert if empty or out of bounds";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: at(key) only reverts if key is out of bounds                                                                  |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule onlyOutOfBoundsRevert(uint256 key) {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    at_@withrevert(key);
+
+    assert lastReverted <=> key >= length(), "only reverts if key is out of bounds";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: only clear/push/pop operations can change the length of the queue                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule noLengthChange(env e) {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    method f;
+    calldataarg args;
+
+    uint256 lengthBefore = length();
+    f(e, args);
+    uint256 lengthAfter = length();
+
+    assert lengthAfter != lengthBefore => (
+        (f.selector == pushFront(bytes32).selector && lengthAfter == lengthBefore + 1) ||
+        (f.selector == pushBack(bytes32).selector && lengthAfter == lengthBefore + 1) ||
+        (f.selector == popBack().selector && lengthAfter == lengthBefore - 1) ||
+        (f.selector == popFront().selector && lengthAfter == lengthBefore - 1) ||
+        (f.selector == clear().selector && lengthAfter == 0)
+    ), "length is only affected by clear/pop/push operations";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: only push/pop can change values bounded in the queue (outters aren't cleared)                                 │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule noDataChange(env e) {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    method f;
+    calldataarg args;
+
+    uint256 key;
+    bytes32 atBefore = at_(key);
+    f(e, args);
+    bytes32 atAfter = at_(key);
+
+    assert atAfter != atBefore => (
+        f.selector == clear().selector || // Although doesn't change values, outters are symbolic, so `* != *`
+        (f.selector == popBack().selector && key == length()) || // at_ only changes if it's left out, becoming symbolic 
+        f.selector == popFront().selector ||
+        f.selector == pushFront(bytes32).selector
+    ), "values of the queue are only changed by a clear, pop or push operation";
```

In some of these cases, like `clear`, the second `at_` shohuld revert. The comments about values being symbolic don't sound right to me, those cases would not be considered for verification.

I think this needs to be something like this:

```suggestion
    bytes32 atAfter = at_@withrevert(key);
    bool atAfterSuccess = !lastReverted;

    assert !atAfterSuccess <=> (
        f.selector == clear().selector ||
        (f.selector == popBack().selector && key == length())
    ), "indexes of the queue are only removed by clear or popBack";

    assert atAfterSuccess && atAfter != atBefore => (
        f.selector == popFront().selector ||
        f.selector == pushFront(bytes32).selector
    ), "values of the queue are only changed by popFront or pushFront";
```

---

The `at_` function is proven to fail iff a key is out of bounds by the `onlyOutOfBoundsRevert` rule and also each `(pop/push)(Front/Back)Consistency` rule makes sure `afterAt == beforeAt`.

In this case, not having the `@withrevert` was on purpose since it guarantees there would be at least one value left, but I think you're right that the case should be accounted.

Regarding your suggestion, I think the first assertion is not checking the same thing, it just checks the implication in the case failure, which is when the queue is left empty, and we want to check that for every `popBack()`

This is the rule originally (the `@withrevert` was incorrectly left out now that I think about it):

```
rule noDataChange(env e) {
    requireInvariant boundariesConsistency();
    require boundedQueue();

    method f;
    calldataarg args;

    uint256 key;
    bytes32 atBefore = at_(key);
    f(e, args);
    bytes32 atAfter = at_@withrevert(key);

    assert atAfter != atBefore => (
        f.selector == clear().selector || // Although doesn't change values, outters are symbolic, so `* != *`
        (f.selector == popBack().selector && key == length()) || // at_ only changes if it's left out, becoming symbolic 
        f.selector == popFront().selector ||
        f.selector == pushFront(bytes32).selector
    ), "values of the queue are only changed by a clear, pop or push operation";
}
```

Having a `@withrevert` there will indeed include the revert cases, but in those cases is when `atAfter` becomes symbolic. I'd rather only add `@withrevert` to the second `at_` as in the snippet above. What do you think?

---

> The `at_` function is proven to fail iff a key is out of bounds by the `onlyOutOfBoundsRevert` rule

I think this is different than that. `onlyOutOfBoundsRevert` is a property of a single `at_` call. Here what I was trying to express is that if an `at_` call succeeds, the only way that the same `at_` call can be made to fail is through `clear` or `popBack` with `key == length`. I think this is valuable information that we lose without `@withrevert`, but it seems we agree that we should add it.

I wasn't sure what the return value of a reverted call would be. Is it documented somewhere that it's a "symbolic" value? I'm not sure what that implies.

---

I guess your proposal also works, but my intuition is that it's weaker than what I proposed, which is also showing exactly when `at(key)` can go from returning to reverting.

---

> In some of these cases, like `clear`, the second `at_` should revert. The comments about values being symbolic don't sound right to me, those cases would not be considered for verification without `@withrevert`.
> 
> I think this needs to be something like this:

👍🏻 

This feels better (if it passes)

---

> What I proposed would show exactly when at(key) can go from returning to reverting.

I see. Since that's the objective, then I added:

```
(f.selector == popFront().selector && key == length())
```

to the first rule, since it's not only popBack reverting.

Sent a commit with your suggestions.